[Expanded Weights] fix layer norm

samdow opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #80944
* #80943
* __->__ #80895
* #80892

Opacus found that Layer Norm can fail from a wrong ordering in the ExpandedWeights code. What was happening is that all our tests had the input require grad so a layer norm check was always short circuiting in the tests, avoiding the wrong ordering. This adds a test where the input does not require gradients and fixes the issue in Layer Norm

Closes #80952

Commit:
fix layer norm

[ghstack-poisoned]

Commit:
Update on "[Expanded Weights] fix layer norm"


Opacus found that Layer Norm can fail from a wrong ordering in the ExpandedWeights code. What was happening is that all our tests had the input require grad so a layer norm check was always short circuiting in the tests, avoiding the wrong ordering. This adds a test where the input does not require gradients and fixes the issue in Layer Norm

[ghstack-poisoned]

Commit:
Update on "[Expanded Weights] fix layer norm"


Opacus found that Layer Norm can fail from a wrong ordering in the ExpandedWeights code. What was happening is that all our tests had the input require grad so a layer norm check was always short circuiting in the tests, avoiding the wrong ordering. This adds a test where the input does not require gradients and fixes the issue in Layer Norm

[ghstack-poisoned]

samdow commented on test/test_expanded_weights.py:
```diff
@@ -193,6 +193,16 @@ def test_expanded_weight_per_sample_grad_mean(self, device, dtype, op):
+        for sample_input in supported_inputs(op, sample_inputs):
+            if op.name == "nn.functional.embedding":  # embedding flips its argument order for autograd tests
+                sample_input = SampleInput(sample_input.args[0], args=(sample_input.input,), kwargs=sample_input.kwargs)
+            sample_input.input.requires_grad_(False)
```
For this, no because that will make all the arguments, including weights, have requires_grad=False. We still want the weights to require gradients so that we test Expanded Weights backwards but we just don't want the input to have a gradient to test the non-short circuited pathway (it checks if input requires gradient since we might be able to avoid saving state if the input doesn't require gradients)

zou3519 replied:
Gotcha, thanks for the explanation

Commit:
Update on "[Expanded Weights] fix layer norm"


Opacus found that Layer Norm can fail from a wrong ordering in the ExpandedWeights code. What was happening is that all our tests had the input require grad so a layer norm check was always short circuiting in the tests, avoiding the wrong ordering. This adds a test where the input does not require gradients and fixes the issue in Layer Norm

Closes #80952

[ghstack-poisoned]

Commit:
Update on "[Expanded Weights] fix layer norm"


Opacus found that Layer Norm can fail from a wrong ordering in the ExpandedWeights code. What was happening is that all our tests had the input require grad so a layer norm check was always short circuiting in the tests, avoiding the wrong ordering. This adds a test where the input does not require gradients and fixes the issue in Layer Norm

Closes #80952

[ghstack-poisoned]

Commit:
Update on "[Expanded Weights] fix layer norm"


Opacus found that Layer Norm can fail from a wrong ordering in the ExpandedWeights code. What was happening is that all our tests had the input require grad so a layer norm check was always short circuiting in the tests, avoiding the wrong ordering. This adds a test where the input does not require gradients and fixes the issue in Layer Norm

Closes #80952

[ghstack-poisoned]

Commit:
Update on "[Expanded Weights] fix layer norm"


Opacus found that Layer Norm can fail from a wrong ordering in the ExpandedWeights code. What was happening is that all our tests had the input require grad so a layer norm check was always short circuiting in the tests, avoiding the wrong ordering. This adds a test where the input does not require gradients and fixes the issue in Layer Norm

Closes #80952

[ghstack-poisoned]

Commit:
Update on "[Expanded Weights] fix layer norm"


Opacus found that Layer Norm can fail from a wrong ordering in the ExpandedWeights code. What was happening is that all our tests had the input require grad so a layer norm check was always short circuiting in the tests, avoiding the wrong ordering. This adds a test where the input does not require gradients and fixes the issue in Layer Norm

Closes #80952

[ghstack-poisoned]

Commit:
Update on "[Expanded Weights] fix layer norm"


Opacus found that Layer Norm can fail from a wrong ordering in the ExpandedWeights code. What was happening is that all our tests had the input require grad so a layer norm check was always short circuiting in the tests, avoiding the wrong ordering. This adds a test where the input does not require gradients and fixes the issue in Layer Norm

Closes #80952

[ghstack-poisoned]
